src/main/java/teammates/ui/webapi/action/GetStudentProfileAction.java

```diff
@@ -1,30 +1,78 @@
 package teammates.ui.webapi.action;
 
+import teammates.common.datatransfer.attributes.InstructorAttributes;
+import teammates.common.datatransfer.attributes.StudentAttributes;
 import teammates.common.datatransfer.attributes.StudentProfileAttributes;
 import teammates.common.exception.UnauthorizedAccessException;
 import teammates.common.util.Const;
-import teammates.ui.webapi.output.ApiOutput;
+import teammates.ui.webapi.output.StudentProfileData;
 
 /**
- * Action: Get a student's profile.
+ * Get a student's profile by an instructor or a classmate of the student.
  */
 public class GetStudentProfileAction extends Action {
+
+    private static final String MESSAGE_NOT_CLASSMATE = "You are not this student's classmate,"
+            + " so you cannot view his profile";
+    private static final String MESSAGE_NOT_TEAMMATE = "You are not this student's teammate,"
+            + " so you cannot view his profile";
+    private static final String MESSAGE_STUDENT_NOT_FOUND = "The student is not in the course you are given,"
+            + " so you cannot access his profile.";
+    private static final String MESSAGE_INVALID_ACCOUNT = "Your account cannot access this student's profile.";
+
     @Override
     protected AuthType getMinAuthLevel() {
         return AuthType.LOGGED_IN;
     }
 
     @Override
     public void checkSpecificAccessControl() {
-        if (!userInfo.isStudent) {
-            throw new UnauthorizedAccessException("Student privilege is required to access this resource.");
+
+        String entityType = getNonNullRequestParamValue(Const.ParamsNames.ENTITY_TYPE);
+        if (userInfo.isStudent && entityType.equals(Const.EntityType.STUDENT)) {
+            checkStudentAccess();
+        } else if (userInfo.isInstructor && entityType.equals(Const.EntityType.INSTRUCTOR)) {
+            checkInstructorAccess();
+        } else {
+            throw new UnauthorizedAccessException(MESSAGE_INVALID_ACCOUNT);
+        }
+    }
+
+    private void checkStudentAccess() {
```

Comment:
I was working on the student profile picture and used a similar method to this (except that it was in Gatekeper: ` gateKeeper.verifyAccessibleForCurrentUserAsInstructorOrTeamMemberOrAdmin`

Perhaps we should add this in Gatekeeper or modify the method I used? This way other Actions can make use of it